File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/games/bridges/android/BridgesGameActivity.java
package com.zwstudio.logicpuzzlesandroid.games.bridges.android;

import com.zwstudio.logicpuzzlesandroid.R;
import com.zwstudio.logicpuzzlesandroid.games.bridges.data.BridgesDocument;
import com.zwstudio.logicpuzzlesandroid.games.bridges.data.BridgesMoveProgress;
import com.zwstudio.logicpuzzlesandroid.games.bridges.domain.BridgesGame;
import com.zwstudio.logicpuzzlesandroid.games.bridges.domain.BridgesGameMove;
import com.zwstudio.logicpuzzlesandroid.games.bridges.domain.BridgesGameState;
import com.zwstudio.logicpuzzlesandroid.common.android.GameActivity;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;

import org.androidannotations.annotations.EActivity;

import java.util.List;

@EActivity(R.layout.activity_bridges_game)
public class BridgesGameActivity extends GameActivity<BridgesGame, BridgesDocument, BridgesGameMove, BridgesGameState> {
    public BridgesDocument doc() {return app.bridgesDocument;}

    protected void startGame() {
        String selectedLevelID = doc().selectedLevelID;
        List<String> layout = doc().levels.get(selectedLevelID);
        tvLevel.setText(selectedLevelID);

        levelInitilizing = true;
        game = new BridgesGame(layout, this);
        try {
            // restore game state
            for (BridgesMoveProgress rec : doc().moveProgress())
                game.switchBridges(new Position(rec.rowFrom, rec.colFrom), new Position(rec.rowTo, rec.colTo));
            int moveIndex = doc().levelProgress().moveIndex;
            if (!(moveIndex >= 0 && moveIndex < game.moveCount())) return;
            while (moveIndex != game.moveIndex())
                game.undo();
        } finally {
            levelInitilizing = false;
        }
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/bridges/BridgesGameActivity.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.bridges

import com.zwstudio.logicpuzzlesandroid.R
import com.zwstudio.logicpuzzlesandroid.common.android.GameGameActivity
import com.zwstudio.logicpuzzlesandroid.common.data.GameLevel
import org.androidannotations.annotations.AfterViews
import org.androidannotations.annotations.Bean
import org.androidannotations.annotations.Click
import org.androidannotations.annotations.EActivity

@EActivity(R.layout.activity_game_game)
class BridgesGameActivity : GameGameActivity<BridgesGame, BridgesDocument, BridgesGameMove, BridgesGameState>() {
    @Bean
    protected lateinit var document: BridgesDocument
    override val doc get() = document

    @AfterViews
    override fun init() {
        gameView = BridgesGameView(this)
        super.init()
    }

    override fun newGame(level: GameLevel) =
        BridgesGame(level.layout, this, doc)

    @Click
    protected fun btnHelp() {
        BridgesHelpActivity_.intent(this).start()
    }
}